### DISCUSSION TITLE
RabbitMQ acknowledges multiple publishes in a single TCP packet

### DISCUSSION THREAD

**Discussion by shaxiaozz** [rabbitmq/rabbitmq-server#10015] [Other] [answered]
### Describe the bug

I'm using a java spring-boot-starter-amqp client for Basic.Publish and I'm in production to push messages to MQ and I need MQ to return acknowledgements. But it often occurs occasionally in my production environment:
```
Caused by: java.util.concurrent.TimeoutException: null
	at com.rabbitmq.client.impl.ChannelN.waitForConfirms(ChannelN.java:229)
	at org.springframework.amqp.rabbit.connection.PublisherCallbackChannelImpl.waitForConfirms(PublisherCallbackChannelImpl.java:665)
	at sun.reflect.GeneratedMethodAccessor845.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.amqp.rabbit.connection.CachingConnectionFactory$CachedChannelInvocationHandler.invoke(CachingConnectionFactory.java:1184)
	at com.sun.proxy.$Proxy313.waitForConfirms(Unknown Source)
	at org.springframework.amqp.rabbit.core.RabbitTemplate.waitForConfirms(RabbitTemplate.java:2241)
	... 35 common frames omitted
```

My RabbitMQ-Server version is 3.7.8 Erlang 21.1

Here are the packets I grabbed from the client to communicate with RabbitMQ:

I found that when there is a waitForConfirms timeout, instead of RabbitMQ replying to the Basic.ACK of the corresponding delivery_tag on this channel, one packet replies to almost 20 different or duplicate delivery_tag Basic.ACKs, and then after 3s (the client waitForConfirms timeout=3s) the client voluntarily closed this connection.

Please pay attention to the place of the red frame, client 10.82.48.237 Basic.Publish production message, and after a packet, RabbitMQ server is the same packet to return more than one Basic.ACK, and after 3s, the client 10.82.49.237 initiates the Channel.Close, thus triggering the waitForConfirms timeout
![image](https://github.com/rabbitmq/rabbitmq-server/assets/43721571/603f5c72-239f-4424-8d7a-e38473438cd5)

I don't know why RabbitMQ responds to the Basic.ACK in this way, ideally RabbitMQ should return the Basic.ACK with delivery_tag=4. Instead of a packet containing multiple Basic.ACKs, I'd like to see it with delivery_tag=1 on top of the packet. Delivery_tag=1, delivery_tag=2, delivery_tag=3 that way operation.

I don't know what the problem is, you guys.

### Reproduction steps

When the client has finished producing messages, and is waiting for RabbitMQ to return ACKs, and RabbitMQ contains multiple ACKs in the same TCP packet level, I will have a waitForConfirms timeout.

### Expected behavior

I don't know why RabbitMQ responds to the Basic.ACK in this way, ideally RabbitMQ should return the Basic.ACK with delivery_tag=4. Instead of a packet containing multiple Basic.ACKs, I'd like to see it with delivery_tag=1 on top of the packet. Delivery_tag=1, delivery_tag=2, delivery_tag=3 that way operation.

**Comment by michaelklishin** [comment, marked as answer]
I assume that by "did not send a single packet" you really mean "multiple acknowledgements were sent in the same TCP packet" because "not send a single packet" means that none acknowledgements were sent and this is not the case.

@shaxiaozz RabbitMQ does not really control how the data it sends is organized into TCP packets. More importantly, acknowledging multiple deliveries at once is a perfectly fine thing to do. AMQP 0-9-1's `basic.ack` frame allows both clients and RabbitMQ to acknowledge multiple deliveries (or published messages) at once.

I don't see why it would "ideally not do that". It makes publisher confirms handling somewhat more complex if you use [streaming confirms](https://rabbitmq.com/publishers.html#data-safety) but some users would say multi-acks is the right thing to do under load.

**Reply by shaxiaozz** [reply]
@michaelklishin Thank you for your reply.
But I still have a question, does RabbitMQ independently choose to acknowledge multiple deliveries at once or just one at a time?

So I have observed many requests with errors, when the producer pushes the message and RabbitMQ receives it, it immediately confirms multiple deliveries at once, and then my client is using this single confirmation, I understand is this the way the client task is not receiving the confirmation message?

**Reply by shaxiaozz** [reply]
@michaelklishin 
But I can see from the data message that multiple is false.
![image](https://github.com/rabbitmq/rabbitmq-server/assets/43721571/6ede891e-2d4a-4bc3-808b-8378c8cda02d)

**Reply by lukebakken** [reply]
Yes, you have several AMQP ack frames in the same TCP packet. Like @michaelklishin said, RabbitMQ / Erlang have **no control** over how the data is organized into TCP packets.

**Reply by michaelklishin** [reply]
It is false for the AMQP 0-9-1 method. You are looking at a TCP packet with multiple AMQP 0-9-1 frames.

The bottom line is: RabbitMQ does not guarantee that acknowledgements will arrive sequentially in order, one by one. Acknowledgements are sent by queues that do not coordinate, and five published messages can be routed to five different queues, producing five concurrent responses, that the kernel will then combine as it sees fit into TCP packets.

When the `multiple` flag is set to true, it will represent a sequential range up to the `delivery_tag` value.

We don't have plans to change this behavior. See how RabbitMQ Java client and PerfTest deal with this, it's a fairly standard situation in networking protocols, including TCP itself (packet can arrive out of order and their sequence is reconstructed by the receiving side).

**Reply by shaxiaozz** [reply]
@michaelklishin I'm very sorry, it is normal behavior for RabbitMQ to have multiple Basic.ACKs in the same TCP packet, I just learned that now. Thanks for informing me.

But I found another problem from the packet capture, the delivery-tag on the same channel is not consecutive, I read the official document description that the value of delivery-tag is counted from 1.
![image](https://github.com/rabbitmq/rabbitmq-server/assets/43721571/9d6991de-8f23-485d-88cb-06c539bc1315)

This puzzled me, I continue to look down, and found that this is very often the case

https://www.rabbitmq.com/confirms.html#publisher-confirms
![image](https://github.com/rabbitmq/rabbitmq-server/assets/43721571/6d245833-b761-4f44-a57e-a8651d8dfad7)

**Reply by lukebakken** [reply]
Please provide the code you are using to publish messages and receive confirmations. I should be able to clone it and run it, so provide instructions if it is not obvious. I don't have time to guess at how to reproduce this.

I also don't think it's a problem. the only guarantee I believe is that the numbers are increasing, not that they are sequential (https://www.rabbitmq.com/amqp-0-9-1-reference.html#basic.ack)

Finally, for delivery tag 13, is `Multiple` set to `true`? It is unclear from your screenshot. My guess is that the `Basic.Ack` for tag 13 is a multi-ack.

**Reply by shaxiaozz** [reply]
@lukebakken I'm very sorry for only getting back to you now.

I checked delivery tag 13 and Multiple is False.
![4fd7c8ff8dd56f90f241d22debe7df0](https://github.com/rabbitmq/rabbitmq-server/assets/43721571/3ae1e131-734b-4024-95b7-acf55658e590)


Also noticed that the serial number is used as a confirm mechanism by the client and server to maintain their own delivery tag values, how to confirm the message properly if it is not consecutive (Multiple is False).

**Reply by shaxiaozz** [reply]
@michaelklishin @lukebakken 
The publisher code is as follows(Publisher ConfirmType is CORRELATED): 

rabbitTemplate.invoke(template -> {
	template.convertAndSend(exchangeName, routingKey, jsonMessage);
	template.waitForConfirms(3000);
	return true;
});

**Reply by lukebakken** [reply]
Please carefully read what I wrote:

> Please provide the code you are using to publish messages and receive confirmations. I should be able to clone it and run it, so provide instructions if it is not obvious. I don't have time to guess at how to reproduce this.

I can't run the code you provided.

The fact is, I don't have time to look at this. If there is a bug, it is in your code. If something isn't working they way you expect, then compare the behavior with a well-known tool, like [PerfTest](https://perftest.rabbitmq.com/). We know RabbitMQ and our official client libraries work correctly with regard to delivery tags - perhaps our documentation needs to be updated to reflect that they may not be exactly sequential.

**Reply by shaxiaozz** [reply]
@lukebakken 
I apologize for the delay~

I'll take some time to pull out the publisher and consumer code and clearly describe how to reproduce the issue

Thanks again for your reply~

**Reply by michaelklishin** [reply]
@shaxiaozz like I said above, many protocols, including TCP, deal with confirmations (whatever exactly is being confirmed in a particular protocol does not matter) arriving out of order. We cover [three different approaches to working with publisher confirms](https://rabbitmq.com/publishers.html#data-safety) in the docs.

**Reply by shaxiaozz** [reply]
@michaelklishin Thank you very much for your response, I will take another hard look at this phenomenon with my team. Thanks for the link.

**Comment by lukebakken** [comment]
@shaxiaozz - here is what I did to observe a basic set of AMQP publishes with confirmations in my local environment:
* Start RabbitMQ 3.12.10 / Erlang 26.1.2
* Start Wireshark - listen to `localhost:5672`
* Run [PerfTest](https://perftest.rabbitmq.com/) with the following arguments
    ```
    --rate 10 --confirm 1 --pmessages 10 --producers 1 --consumers 0
    ```
* Here is the capture: [localhost-rabbitmq-3.12-perftest-capture.zip](https://github.com/rabbitmq/rabbitmq-server/files/13561958/localhost-rabbitmq-3.12-perftest-capture.zip)

You will note that the delivery tags are sequential.

Please take the time to run the same PerfTest command that I ran in your environment.

I just noticed this in your original message:

> My RabbitMQ-Server version is 3.7.8 Erlang 21.1

RabbitMQ 3.7.8 is OVER FIVE YEARS OLD, and is unsupported. I will _not_ spend any more time on this issue until you update to the latest version of RabbitMQ.

Do not take the time to provide code if you are unable to upgrade RabbitMQ.

**Reply by shaxiaozz** [reply]
@lukebakken Okay, copy that.Thanks.
